### PR TITLE
osd, RFC: make the Session unaware about Watch.

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/Watch.cc [hunk 1]
```diff
@@ -369,17 +369,16 @@ void Watch::got_ping(utime_t t)
...
   dout(10) << __func__ << " con " << con << dendl;
   conn = con;
   auto priv = con->get_priv();
   if (priv) {
     for (auto i = in_progress_notifies.begin();
 	 i != in_progress_notifies.end();
 	 ++i) {
+      // try to send regardless whether we were already connected or not.
+      // this can lead to duplicating the `MWatchNotify` which should be
+      // harmless.
```
Comment (rzarzynski) [inline src/osd/Watch.cc:381]: This claim requires confirmation.

**Comment by rzarzynski** [comment]
Added the _crimson_ label as the PR has consequences for the crimson's n:m mapping.

**Comment by jdurgin** [comment]
I think this should be safe, it should run through the rgw/rbd suites as well since they rely on watch/notify. Added others to verify as well.

**Review by adamemerson — APPROVED** [review]
This is pretty good. I like it.

**Comment by rzarzynski** [comment]
Taking the _DNM_ label out with the hope that teuthology testing will be fine.

**Review by athanatos — APPROVED** [review]
Way simpler.

**Comment by tchaikov** [comment]
tests:

- https://pulpito.ceph.com/kchai-2021-04-15_08:31:03-rados-wip-kefu-testing-2021-04-15-1359-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-16_02:02:40-rados-wip-kefu-testing-2021-04-15-1359-distro-basic-gibba/

failures:

- the test_mon_config_key.py failure is tracked by https://tracker.ceph.com/issues/50393
- the SELinux denials are tracked by https://tracker.ceph.com/issues/50151
- the upgrade issues are tracked by https://tracker.ceph.com/issues/50388
- the valgrind issue is tracked by https://tracker.ceph.com/issues/50299
- the test_pause is caused by WARNING of

  ```
  pgmap v56: 97 pgs: 13 active+undersized+degraded, 84 active+undersized; 22 KiB data, 37 MiB used, 252 GiB / 252 GiB avail; 22/66 objects degraded (33.333%)
  ```
  i will rerun the failed test without this change.

**Review by tchaikov — CHANGES_REQUESTED** [review]
```
2021-04-24T04:44:39.175 ERROR:teuthology.run_tasks:Manager failed: ceph
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_e14b06116d158818e73cedb71cf696d989b153a3/qa/tasks/ceph.py", line 1895, in task
    yield
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_2713a3cd31b17738a50039eaa9d859b5dc39fb8a/teuthology/run_tasks.py", line 176, in run_tasks
    suppress = manager.__exit__(*exc_info)
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_e14b06116d158818e73cedb71cf696d989b153a3/qa/tasks/watch_notify_same_primary.py", line 126, in task
    assert got1 and got2
AssertionError
```

https://pulpito.ceph.com/kchai-2021-04-24_04:07:09-rados-wip-kefu-testing-2021-04-23-2026-distro-basic-smithi/6070010/

**Comment by gregsfortytwo** [comment]
Ping @idryomov, looks like we added the new op in late 2014 but I don't know what that means for the kernel clients we want to support?

**Review by idryomov — CHANGES_REQUESTED** [review]
Sorry, I saw that there were some test failures and the `DNM` label and haven't looked deeper.  Dropping the legacy watch op would effectively break krbd clients on kernels < 4.7 (released in July 2016).  We have a lot of backwards compatibility cruft that ensures that we are compatible far beyond that -- in fact a 3.10ish kernel would still work with old tunables, no msgr2, etc.  @rzarzynski is it critical for crimson n:m mapping or just a nice to have?

Making as "Request changes" because the 4.7 kernel cut off would set a precedent.

**Comment by ljflores** [comment]
> Sorry, I saw that there were some test failures and the `DNM` label and haven't looked deeper. Dropping the legacy watch op would effectively break krbd clients on kernels < 4.7 (released in July 2016). We have a lot of backwards compatibility cruft that ensures that we are compatible far beyond that -- in fact a 3.10ish kernel would still work with old tunables, no msgr2, etc. @rzarzynski is it critical for crimson n:m mapping or just a nice to have?
> 
> Making as "Request changes" because the 4.7 kernel cut off would set a precedent.

@rzarzynski ?

**Pull request closed without merging**